crates/dbsp/src/algebra/decimal.rs

```diff
@@ -0,0 +1,607 @@
+
+    impl TryFrom<u32> for Precision {
+        type Error = String;
+
+        fn try_from(value: u32) -> Result<Self, Self::Error> {
+            match value {
+                x if (MIN_DECIMAL_PRECISION..=MAX_DECIMAL_PRECISION).contains(&x) => Ok(Precision(x)),
+                _ => Err(format!(
+                    "Numeric precision {value} must be between {MIN_DECIMAL_PRECISION} and {MAX_DECIMAL_PRECISION}"
+                )),
+            }
+        }
+    }
+
+    impl Display for Precision {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{}", self.0)
+        }
+    }
+
+    impl Default for Precision {
+        fn default() -> Self {
+            Precision(MAX_DECIMAL_PRECISION)
+        }
+    }
+
+    // In Postgres, the scale can be between [-1000, 1000]
+    #[derive(
+        Debug,
+        Clone,
+        Copy,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Hash,
+        SizeOf,
+        rkyv::Archive,
+        rkyv::Serialize,
+        rkyv::Deserialize,
+    )]
+    #[cfg_attr(feature = "with-serde", derive(serde::Serialize, serde::Deserialize))]
+    pub struct Scale(u32);
+
+    impl Scale {
+        pub fn inner(&self) -> u32 {
+            self.0
+        }
+
+        pub fn min() -> Self {
+            Self(MIN_DECIMAL_SCALE)
+        }
+
+        pub const fn max() -> Self {
+            Self(MAX_DECIMAL_SCALE)
+        }
+
+        pub fn apply(&self, val: &mut Decimal) {
+            val.rescale(self.0)
+        }
+    }
+
+    impl TryFrom<u32> for Scale {
+        type Error = String;
+
+        fn try_from(value: u32) -> Result<Self, Self::Error> {
+            match value {
+                x if (MIN_DECIMAL_SCALE..=MAX_DECIMAL_SCALE).contains(&x) => Ok(Scale(x)),
+                _ => Err(format!(
+                "Numeric scale {value} must be between {MIN_DECIMAL_SCALE} and {MAX_DECIMAL_SCALE}"
+            )),
+            }
+        }
+    }
+
+    impl Display for Scale {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            write!(f, "{}", self.0)
+        }
+    }
+
+    impl Default for Scale {
+        fn default() -> Self {
+            Self(MAX_DECIMAL_SCALE)
+        }
+    }
+}
+
+use inner_types::*;
+
+/// A wrapper type around [`rust_decimal::Decimal`].
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, SizeOf)]
+#[cfg_attr(feature = "with-serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "with-serde", serde(transparent))]
+#[derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)]
+pub struct SQLDecimal {
+    inner: Decimal,
+    #[cfg_attr(feature = "with-serde", serde(skip, default))]
+    max_precision: Precision,
+    #[cfg_attr(feature = "with-serde", serde(skip, default))]
+    max_scale: Scale,
+}
+
+impl FromStr for SQLDecimal {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```

Comment:
Is this method ever used? Again, the compiler should always tell you the precision and scale, you should never need to infer them.

Reply:
This method exists because I was thinking of using it for cast to `DECIMAL` and using `from_str_with_precision_scale` for `DECIMAL(x,y)`. But now it can be removed.